### PR TITLE
MLv2: Simplify MetadataProvider protocols (16 methods/3 protocols => 7 methods/2 protocols)

### PR DESCRIPTION
**Pull request by camsaul**
- Consolidate the 12 methods in the two `MetadataProvider` and `BulkMetadataProvider` protocols into 5 methods in a single protocol -- instead of separate methods to fetch a Field/Card/Table/LegacyMetric/Segment as well as a bulk method, just have the metadata providers implement only the bulk method and add helper functions for fetching single objects of different types
- Remove unused `cached-database` and `store-database!` methods from the `CachedMetadataProvider` protocol
- Finally finish the cleanup of moving everything away from using vars like `lib.metadata/LegacyMetricMetadata` to the schema keywords e.g. `::lib.schema.metadata/legacy-metric`

### CODE REVIEW

File: src/metabase/lib/schema/metadata.cljc [hunk 1]
```diff
@@ -256,6 +268,10 @@
    [:description {:optional true} [:maybe ::lib.schema.common/non-blank-string]]])
 
 (mr/def ::legacy-metric
+  "Malli schema for a legacy v1 [[metabase.models.legacy-metric]], but with kebab-case keys. A Metric defines an MBQL
+  snippet with an aggregation and optionally a filter clause. You can add a `:metric` reference to the `:aggregations`
```
Comment (bshepherdson) [inline src/metabase/lib/schema/metadata.cljc:272]: Should that be `:legacy-metric`?

File: src/metabase/models/segment.clj [hunk 2]
```diff
@@ -86,37 +86,37 @@
...
-        fields            (lib.metadata.protocols/bulk-metadata metadata-provider :metadata/column field-ids)
+        fields            (lib.metadata.protocols/metadatas metadata-provider :metadata/column field-ids)
         table-ids         (into #{}
                                 cat
                                 [(map :table-id fields)
                                  (map :table_id segments)])]
     ;; this is done for side effects
-    (lib.metadata.protocols/bulk-metadata metadata-provider :metadata/table table-ids)
+    (when (lib.metadata.protocols/cached-metadata-provider? metadata-provider)
+      (lib.metadata.protocols/metadatas metadata-provider :metadata/table table-ids))
```
Comment (bshepherdson) [inline src/metabase/models/segment.clj:104]: This is a bit ugly to see outside of `metabase.lib.metadata.*`.

How about `(lib.metadata/warm-cache metadata-provider :metadata/table table-ids)` (or in `lib.metadata.protocols` if that's a better fit) and let that understand about `cached-metadata-provider?` etc. and do nothing if there's no cache.

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/lib/schema/metadata.cljc (hunk 1 above)
Comment (camsaul) [reply src/metabase/lib/schema/metadata.cljc:272]: It's talking about `:metric` clauses which are still called `:metric` AFAIK even if they are V1 metrics

File: src/metabase/models/segment.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/models/segment.clj:104]: Yeah that's a good point

Comment (camsaul) [reply src/metabase/models/segment.clj]: Implemented this

**Pull request merged**